// connect-mongo/build/main/test/integration.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const supertest_1 = __importDefault(require("supertest"));
const express_1 = __importDefault(require("express"));
const express_session_1 = __importDefault(require("express-session"));
const __1 = __importDefault(require("../"));
function createSupertetAgent(sessionOpts, mongoStoreOpts) {
    const app = (0, express_1.default)();
    const store = __1.default.create(mongoStoreOpts);
    app.use((0, express_session_1.default)({
        ...sessionOpts,
        store: store,
    }));
    app.get('/', function (req, res) {
        if (typeof req.session.views === 'number') {
            req.session.views++;
        }
        else {
            req.session.views = 0;
        }
        res.status(200).send({ views: req.session.views });
    });
    app.get('/ping', function (req, res) {
        res.status(200).send({ views: req.session.views });
    });
    const agent = supertest_1.default.agent(app);
    return agent;
}
function createSupertetAgentWithDefault(sessionOpts = {}, mongoStoreOpts = {}) {
    return createSupertetAgent({ secret: 'foo', ...sessionOpts }, {
        mongoUrl: 'mongodb://root:example@127.0.0.1:27017',
        dbName: 'itegration-test-db',
        stringify: false,
        ...mongoStoreOpts,
    });
}
ava_1.default.serial.cb('simple case', (t) => {
    const agent = createSupertetAgentWithDefault();
    agent
        .get('/')
        .expect(200)
        .then((response) => response.headers['set-cookie'])
        .then((cookie) => {
        agent
            .get('/')
            .expect(200)
            .end((err, res) => {
            t.is(err, null);
            t.deepEqual(res.body, { views: 1 });
            return t.end();
        });
    });
});
ava_1.default.serial.cb('simple case with touch after', (t) => {
    const agent = createSupertetAgentWithDefault({ resave: false, saveUninitialized: false, rolling: true }, { touchAfter: 1 });
    agent
        .get('/')
        .expect(200)
        .then(() => {
        agent
            .get('/')
            .expect(200)
            .end((err, res) => {
            t.is(err, null);
            t.deepEqual(res.body, { views: 1 });
            new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 1200);
            }).then(() => {
                agent
                    .get('/ping')
                    .expect(200)
                    .end((err, res) => {
                    t.is(err, null);
                    t.deepEqual(res.body, { views: 1 });
                    return t.end();
                });
            });
        });
    });
});
