// connect-mongo/build/main/test/testHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStoreHelper = exports.makeDataNoCookie = exports.makeData = exports.makeCookie = void 0;
// eslint-disable-next-line eslint-comments/disable-enable-pair
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
const util_1 = __importDefault(require("util"));
const express_session_1 = __importDefault(require("express-session"));
const MongoStore_1 = __importDefault(require("../lib/MongoStore"));
// Create a connect cookie instance
const makeCookie = () => {
    const cookie = new express_session_1.default.Cookie();
    cookie.maxAge = 10000; // This sets cookie.expire through a setter
    cookie.secure = true;
    cookie.domain = 'cow.com';
    cookie.sameSite = false;
    return cookie;
};
exports.makeCookie = makeCookie;
// Create session data
const makeData = () => {
    return {
        foo: 'bar',
        baz: {
            cow: 'moo',
            chicken: 'cluck',
        },
        num: 1,
        cookie: (0, exports.makeCookie)(),
    };
};
exports.makeData = makeData;
const makeDataNoCookie = () => {
    return {
        foo: 'bar',
        baz: {
            cow: 'moo',
            fish: 'blub',
            fox: 'nobody knows!',
        },
        num: 2,
    };
};
exports.makeDataNoCookie = makeDataNoCookie;
const createStoreHelper = (opt = {}) => {
    const store = MongoStore_1.default.create({
        mongoUrl: 'mongodb://root:example@127.0.0.1:27017',
        mongoOptions: {},
        dbName: 'testDb',
        collectionName: 'test-collection',
        ...opt,
    });
    const storePromise = {
        length: util_1.default.promisify(store.length).bind(store),
        clear: util_1.default.promisify(store.clear).bind(store),
        get: util_1.default.promisify(store.get).bind(store),
        set: util_1.default.promisify(store.set).bind(store),
        all: util_1.default.promisify(store.all).bind(store),
        touch: util_1.default.promisify(store.touch).bind(store),
        destroy: util_1.default.promisify(store.destroy).bind(store),
        close: store.close.bind(store),
    };
    return { store, storePromise };
};
exports.createStoreHelper = createStoreHelper;
